libraries/esp8266/examples/UartDownload/UartDownload.ino

```diff
@@ -0,0 +1,181 @@
+constexpr size_t kSyncTimeoutMs = 10;
+
+// The SLIP Frame end character, which is also used to start a frame.
+constexpr char slipFrameMarker = '\xC0';
+
+// General packet format:
+//   <0xC0><cmd><payload length><32 bit cksum><payload data ...><0xC0>
+// Slip packet for ESP_SYNC, minus the frame markers ('\xC0').
+// Captured from esptool using the `--trace` option.
+const char syncPkt[] PROGMEM =
+  "\x00\x08\x24\x00\x00\x00\x00\x00\x07\x07\x12\x20"
+  "UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU";
+
+constexpr size_t syncPktSz = sizeof(syncPkt) - 1; // Don't compare zero terminator char
+
+//
+//  Use the discovery of an ESP_SYNC packet, to trigger calling UART Download Mode.
+//
+void proxyEspSync() {
+
+  if (!uartDownloadEnable ||
+      0 >= Serial.available() ||
+      slipFrameMarker != Serial.peek()) {
+    return;
+  }
+
+  Serial.read();      // Skip (start) Frame marker
+
+  byte buf[pktBufSz];
+
+  // If it is an ESP_SYNC packet, it will not take long for readBytesUntil() to
+  // complete.
+  Serial.setTimeout(kSyncTimeoutMs);
+  int len = Serial.readBytesUntil(slipFrameMarker, buf, pktBufSz);
+
+  // To avoid a false trigger, only start UART Download Mode when we get an
+  // exact match to the captured esptool ESP_SYNC packet.
+  if (syncPktSz == len && 0 == memcmp_P(buf, syncPkt, len)) {
+    ESP.rebootIntoUartDownloadMode();
+    // Does not return
+  }
+
+  // Assume RX FIFO data is garbled and flush all RX data.
+  while (0 <= Serial.read()) {} // Clear FIFO
+}
+//
+////////////////////////////////////////////////////////////////////////////////
+
+void preinit() {
+  // (no C++ in function)
+  // disable wifi
+  ESP8266WiFiClass::preinitWiFiOff();
+}
+
+void setup() {
+  // For `proxyEspSync()` to work, the Serial.begin() speed needs to be
+  // 115200bps. This is the data rate used by esptool.py. It expects the Boot
+  // ROM to use its "auto-baud" feature to match up. Since `proxyEspSync()` is
+  // acting as a proxy we must use 115200.
+  //
+  // If on the Arduino IDE Tools menu you use "Upload Speeds" above 115200, it
+  // will work. When esptool.py is run with the `--baud BAUD` option specified
+  // above 115200, initial communication with the ESP8266 is done at 115200bps.
+  // Once esptool.py has synchronize with the ESP8266 and downloaded a short
+  // stub, then both devices shift their UART speeds to the command line value.
+  Serial.begin(115200);
+
+  Serial.println(F(
+                   "\r\n\r\n"
+                   "Boot UART Download Demo - initialization started.\r\n"
+                   "\r\n"
+                   "For a quick test to see the UART Download work,\r\n"
+                   "stop your serial terminal APP and run:\r\n"
+                   "  esptool.py --chip esp8266 --before no_reset --after soft_reset flash_id\r\n"));
+
+  // ...
+}
+
+void cmdLoop(Stream& oStream) {
+  if (0 >= oStream.available()) {
+    return;
+  }
+
+  switch (oStream.read()) {
```

Comment:
This might eat the `slipFrameMarker`. May be also call `proxyEspSync()` in case of `slipFrameMarker`.

Beside this LGTM.

Reply:
Yep, race condition. I tried to separate the two operations in `loop()` for presentation and they really need to be integrated.

Reply:
Would the logic be simplified if `oStream.peek()` is used instead ?

Reply:
I refactored this earlier this morning. I pushed my changes. Let me know what you think.